<!doctype html>
<html class="no-js" lang="en">

<head>
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-134931760-1"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-134931760-1');
  </script>

  <meta charset="utf-8">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <title>Centralized exception handling in Node.JS</title>
  <link rel="icon" type="image/x-icon" href="/favicon.ico" />
  <meta name="description" content="Preamble from the translator: a couple of months ago I was looking for a solution to be able to use exceptions in the game server, written in node.js....">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <link rel="stylesheet" href="../../css/main.css">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script>window.jQuery || document.write('<script src="../../js/vendors/jquery-3.3.1.min.js"><\/script>')</script>

  <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script>
       (adsbygoogle = window.adsbygoogle || []).push({
            google_ad_client: "ca-pub-6974184241884155",
            enable_page_level_ads: true
       });
  </script>
</head>

<body>
  <!--[if lte IE 9]>
    <p class="browserupgrade">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience and security.</p>
  <![endif]-->
  <header class="page-header js-page-header">
    <a class="page-header-logo-container" href="https://weekly-geekly.github.io/index.html"></a>
    <div class="page-header-text">Geekly Articles each Day</div>
  </header>
  <nav class="page-headings-container js-page-headings-container"></nav>
  <div class="tools-bar js-tools-bar">
    <!-- <a href="../../search.html" title="Search">üîé</a> -->
    <a class="js-list-of-headings-button" data-state="closed" href="#" title="Headings">üìú</a>
    <a class="js-go-to-top-button" href="#" title="Go to Top">‚¨ÜÔ∏è</a>
    <a class="js-go-to-bottom-button" href="#" title="Go to Bottom">‚¨áÔ∏è</a>
  </div>
  <a href="http://bit.ly/donateToWeeklyGeekly" class="donate-btn">DONATE</a>
  <section class="page js-page"><h1>Centralized exception handling in Node.JS</h1><div class="post__text post__text-html js-mediator-article"><img src="https://habrastorage.org/storage2/23c/c82/119/23cc8211926b7b36a2c32f64a47644bd.jpg"><br><br>  <i>Preamble from the translator: a couple of months ago I was looking for a solution to be able to use exceptions in the game server, written in node.js.</i>  <i>Unfortunately, pure exceptions are not quite compatible with the environment running on the event loop.</i>  <i>The easiest way to explain this is by example:</i> <br><pre><code class="javascript hljs"><span class="hljs-keyword"><span class="hljs-keyword">try</span></span> { process.nextTick(<span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"></span><span class="hljs-function"><span class="hljs-params"></span>) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> <span class="hljs-built_in"><span class="hljs-built_in">Error</span></span>(<span class="hljs-string"><span class="hljs-string">'Catch Me If You Can'</span></span>); }); } <span class="hljs-keyword"><span class="hljs-keyword">catch</span></span> (e) { <span class="hljs-built_in"><span class="hljs-built_in">console</span></span>.log(<span class="hljs-string"><span class="hljs-string">'Exception caught:'</span></span>, e); }</code> </pre> <br>  <i>This exception, of course, will not be caught, and it will drop the whole process.</i>  <i>A month ago, node.js version 0.8.0 saw the light with the fresh (experimental) <a href="http://nodejs.org/api/domain.html">domain</a> module, which is designed to solve such problems.</i>  <i>However, I would like to pay tribute to the class that I still use.</i>  <i>Go:</i> <br><a name="habracut"></a><br>  Functional programming in node.js is fun, expressive and compact.  Except for one thing - exception handling.  This is not often said, but in my opinion, the lack of a harmonious way of handling errors and exceptions is one of the biggest drawbacks of node.js.  <a href="https://github.com/laverdet/node-fibers/">Node-fibers</a> uses a fully imperative programming style to achieve this, but I would like to solve a problem within a functional style. <br><br>  The problem with error codes (on which the node.js core is based) is that the code that encounters the error first is almost always not the place to decide how to react to it.  In this case, try / catch structures in multi-threaded systems are much more understandable.  Someone higher in stack usually knows how to handle the error. 
      <br>
        <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
        <ins class="adsbygoogle"
          style="display:block; text-align:center;"
          data-ad-layout="in-article"
          data-ad-format="fluid"
          data-ad-client="ca-pub-6974184241884155"
          data-ad-slot="8945601208"></ins>
        <script>
          (adsbygoogle = window.adsbygoogle || []).push({});
        </script>
      <br>
      However, the problem with asynchronous systems, such as node, is that every time one of your callbacks or EventEmitters is called, it is either called at the very top of the event loop, or it is caused by code other than this listener (the place where we assign the listener is perhaps a better place to handle a potential error than some arbitrary code that caused it).  If you throw an exception in such conditions, your program will most likely fall entirely.  Considering that there are so many possibilities in JavaScript, when the right code can throw out runtime errors, this problem becomes even worse than in C, where I can avoid problems if I‚Äôm careful with pointers and won‚Äôt divide by zero.  Yes, unit tests help, but it‚Äôs more like trying to plug all holes in the sieve when you really just need a bowl. <br><br>  To achieve this, we need the ability to set a Block with our own Error Handler, which can be created quickly and easily, and passed along with callbacks to an external code.  Then, if an exception occurs in the callback, it should be sent to the Block Error Handler that was active when we created the callback.  I found that most of these solutions are introduced by Futures, Promises, Fibers, etc.  along with this simple functionality for creating Blocks.  The following snippet describes the Block class, which does exactly what I need: <br><pre> <code class="javascript hljs"><span class="hljs-comment"><span class="hljs-comment">/** *  Block        . */</span></span> <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">Block</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">errback</span></span></span><span class="hljs-function">) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">this</span></span>._parent=Block.current; <span class="hljs-keyword"><span class="hljs-keyword">this</span></span>._errback=errback; } Block.current=<span class="hljs-literal"><span class="hljs-literal">null</span></span>; <span class="hljs-comment"><span class="hljs-comment">/** *  ,   ,     , *           guard(). *     ,   . * * Example: stream.on('end', Block.guard(function() { ... })); */</span></span> Block.guard=<span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">f</span></span></span><span class="hljs-function">) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">if</span></span> (<span class="hljs-keyword"><span class="hljs-keyword">this</span></span>.current) <span class="hljs-keyword"><span class="hljs-keyword">return</span></span> <span class="hljs-keyword"><span class="hljs-keyword">this</span></span>.current.guard(f); <span class="hljs-keyword"><span class="hljs-keyword">else</span></span> <span class="hljs-keyword"><span class="hljs-keyword">return</span></span> f; }; <span class="hljs-comment"><span class="hljs-comment">/** *      .  -     (" try"). *  -    ('catch'). */</span></span> Block.begin=<span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">block, rescue</span></span></span><span class="hljs-function">) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">var</span></span> ec=<span class="hljs-keyword"><span class="hljs-keyword">new</span></span> Block(rescue); <span class="hljs-keyword"><span class="hljs-keyword">return</span></span> ec.trap(block); }; <span class="hljs-comment"><span class="hljs-comment">/** *   function(err),        , *       (  ,    ). *      err  true. * * Example: request.on('error', Block.errorHandler()) */</span></span> Block.errorHandler=<span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"></span><span class="hljs-function"><span class="hljs-params"></span>) </span></span>{ <span class="hljs-comment"><span class="hljs-comment">// Capture the now current Block for later var current=this.current; return function(err) { if (!err) return; if (current) return current.raise(err); else throw err; }; }; /** *    .     error handler,     . *   raise(...)     .    , *        throw. *  ,   error handler'  , *     error handler'. */ Block.prototype.raise=function(err) { if (this._errback) { try { this._errback(err); } catch (nestedE) { if (this._parent) this._parent.raise(nestedE); else throw nestedE; } } else { if (this._parent) this._parent.raise(err); else throw(err); } }; /** *   callback    . *       raise()  . *  return value   undefined   . */ Block.prototype.trap=function(callback) { var origCurrent=Block.current; Block.current=this; try { var ret=callback(); Block.current=origCurrent; return ret; } catch (e) { Block.current=origCurrent; this.raise(e); } }; /** *  ,        . *     trap(),       . */ Block.prototype.guard=function(f) { if (f.__guarded__) return f; var self=this; var wrapped=function() { var origCurrent=Block.current; Block.current=self; try { var ret=f.apply(this, arguments); Block.current=origCurrent; return ret; } catch (e) { Block.current=origCurrent; self.raise(e); } }; wrapped.__guarded__=true; return wrapped; };</span></span></code> </pre><br><br>  (I chose the terminology of Block / Rescue, not because I feel affection for Ruby, but because such a solution does not use the words reserved in JS). <br><br>  <i>Note</i>  <i>translator: an example from the beginning of the article, but using Blocks, takes the following form:</i> <br><pre> <code class="javascript hljs">Block.begin(<span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"></span><span class="hljs-function"><span class="hljs-params"></span>) </span></span>{ process.nextTick(Block.guard(<span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"></span><span class="hljs-function"><span class="hljs-params"></span>) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">throw</span></span> <span class="hljs-keyword"><span class="hljs-keyword">new</span></span> <span class="hljs-built_in"><span class="hljs-built_in">Error</span></span>; })); }, <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">err</span></span></span><span class="hljs-function">) </span></span>{ <span class="hljs-built_in"><span class="hljs-built_in">console</span></span>.log(<span class="hljs-string"><span class="hljs-string">'Exception caught:'</span></span>, err); });</code> </pre>  <i>Now the exception is processed, and we can not drop the server.</i>  <i>This also works with setTimeout, EventEmitter, callbacks for requests to the database, and anything else.</i> <br><br>  Now consider an example of using the Block for centralized error handling.  In our example, <a href="https://github.com/senchalabs/connect/">connect</a> 'ovsky middleware is used, in which case the next function is an excellent error handler: it will give the correct error to the http client.  If we had to somehow handle the error ourselves, we could simply describe the callback in the form <code>function(err) { ; next(err); }</code> <code>function(err) { ; next(err); }</code>  <code>function(err) { ; next(err); }</code> .  You can also use inline functions in Block.begin calls for greater visual similarity to try / catch, but I prefer to use named callbacks to increase readability. <br><br><pre> <code class="javascript hljs"><span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">handleUserAgent</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">req, res, next</span></span></span><span class="hljs-function">) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">return</span></span> Block.begin(process, next); <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function"> </span><span class="hljs-title"><span class="hljs-function"><span class="hljs-title">process</span></span></span><span class="hljs-function">(</span><span class="hljs-params"></span><span class="hljs-function"><span class="hljs-params"></span>) </span></span>{ jsonifyRequest(req, withRequest); <span class="hljs-comment"><span class="hljs-comment">//  ,   jsonifyRequest , //    Block.guard() } function withRequest(requestObj) { var r=validators.UserAgentRecord(requestObj, {fix:true}); if (!r.valid) { res.writeHead(400); return res.end('Invalid request object: ' + r.reason); } var uar=r.object; if (uar.token) { // Verify //return handler.verifyUserAgent(uar); throw new Error('verifyUserAgent not yet implemented'); } else { // Create uar.token=null; uar.type='auth'; // </span><span class="hljs-doctag"><span class="hljs-comment"><span class="hljs-doctag">TODO:</span></span></span><span class="hljs-comment"> Maybe support unauth in the future? handler.createUserAgent(uar, Block.guard(withUserAgent)); } } function withUserAgent(userAgent) { var r=validators.UserAgentRecord(userAgent, {fix:true}); return respondJson(r.object, res); } }</span></span></code> </pre><br><br>  The main point to keep in mind is that any exception thrown by this code or the process () function will be forwarded to the error handler (in this case, to the next function).  In order to bind the callbacks to the block, they must be wrapped using Block.guard (originalFunction).  This will remember the active block at the time of calling Block.guard (), and then restore it as a context before calling the originalFunction () function itself. <br><br>  Consider another example of the explicit use of a block in callbacks.  In this case, we make an HTTP request, accumulate the response text, call the callback, passing the created CouchResponse (which parses the response and other things that can throw an exception) into it. <br><br><pre> <code class="javascript hljs">request: <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">options, callback</span></span></span><span class="hljs-function">) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">var</span></span> req=http.request(options, <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">res</span></span></span><span class="hljs-function">) </span></span>{ <span class="hljs-keyword"><span class="hljs-keyword">var</span></span> text=<span class="hljs-string"><span class="hljs-string">''</span></span>; res.setEncoding(<span class="hljs-string"><span class="hljs-string">'utf8'</span></span>); res.on(<span class="hljs-string"><span class="hljs-string">'data'</span></span>, <span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"><span class="hljs-function"><span class="hljs-params">chunk</span></span></span><span class="hljs-function">) </span></span>{ text+=chunk; }); res.on(<span class="hljs-string"><span class="hljs-string">'end'</span></span>, Block.guard(<span class="hljs-function"><span class="hljs-keyword"><span class="hljs-function"><span class="hljs-keyword">function</span></span></span><span class="hljs-function">(</span><span class="hljs-params"></span><span class="hljs-function"><span class="hljs-params"></span>) </span></span>{ callback(<span class="hljs-keyword"><span class="hljs-keyword">new</span></span> CouchResponse(res, text)); })); res.on(<span class="hljs-string"><span class="hljs-string">'error'</span></span>, Block.errorHandler()); }); req.on(<span class="hljs-string"><span class="hljs-string">'error'</span></span>, Block.errorHandler()); req.end(); }</code> </pre><br><br>  There are still several places where we might encounter unexpected exceptions dropping the whole process: <br><ul><li>  Directly in the callback function (res). </li><li>  In the 'data' callback. </li></ul><br>  I could also wrap them using Block.guard, but I think this is redundant.  In addition, I am 100% sure that the error in this case is critical, and should be covered by unit tests.  The 'end' handler, however, does something that I cannot see immediately (and I know that it contains a JSON.parse call), so I prefer to protect it with guard.  Finally, I use the standard errorHandler () block to catch request and response error events.  This simple template for centralized error handling makes it pretty clear where these errors go and handle them at any level where it makes sense.  You can use nested calls to Block.begin () (analogous to <code>try{try{}catch{}}catch{}</code> ).  This is useful in code for frameworks that should do some work within a block created by someone else's code. <br><br>  <i>PS: The author has an implementation of Future and examples with their use.</i>  <i>I did not translate everything related to Future, and adapted the examples for the use of classical callbacks.</i> <br><br>  <i>I recommend to read the original text in its entirety, because</i>  <i>there are as many as 10 recommendations for writing bulletproof code on node.js.</i> <br><br>  <i>I designed the Block class as a <a href="https://github.com/AlexeyKupershtokh/control-block">github repository</a> and an <a href="http://search.npmjs.org/">npm module</a> (npm install control-block).</i> </div><p>Source: <a href="https://habr.com/ru/post/147233/">https://habr.com/ru/post/147233/</a></p>
<section class="more-articles-navigation-panel js-more-articles-navigation-panel">
<h4>More articles:</h4>
<nav class="list-of-articles-container js-list-of-articles-container"><ul class="list-of-pages js-list-of-pages">
<li><a href="../147226/index.html">Font anatomy</a></li>
<li><a href="../147229/index.html">Yota CEO Denis Sverdlov appointed Deputy Minister of Communications</a></li>
<li><a href="../147230/index.html">Transferring Files from a Siemens Phone to Linux - SieFS</a></li>
<li><a href="../147231/index.html">Hi from 1996: personal computer for 5 million rubles</a></li>
<li><a href="../147232/index.html">Editing and pricing Windows Server 2012</a></li>
<li><a href="../147238/index.html">Bash: autodetection of the enemy for online play on Macs</a></li>
<li><a href="../147239/index.html">10 rules that will help the application to receive a recommendation from Google Play</a></li>
<li><a href="../14724/index.html">Star Wars X-Wing Fighter Model Prepares for First Flight</a></li>
<li><a href="../147241/index.html">Generation of alignment for card games</a></li>
<li><a href="../147242/index.html">Blocking mobile parasites</a></li>
</ul></nav>
</section><br />
<a href="../../allArticles.html"><strong>All Articles</strong></a>
<script src="../../js/main.js"></script>

<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
  (function (d, w, c) {
      (w[c] = w[c] || []).push(function() {
          try {
              w.yaCounter52496797 = new Ya.Metrika({
                  id:52496797,
                  clickmap:true,
                  trackLinks:true,
                  accurateTrackBounce:true,
                  webvisor:true
              });
          } catch(e) { }
      });

      var n = d.getElementsByTagName("script")[0],
          s = d.createElement("script"),
          f = function () { n.parentNode.insertBefore(s, n); };
      s.type = "text/javascript";
      s.async = true;
      s.src = "https://mc.yandex.ru/metrika/watch.js";

      if (w.opera == "[object Opera]") {
          d.addEventListener("DOMContentLoaded", f, false);
      } else { f(); }
  })(document, window, "yandex_metrika_callbacks");
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/52496797" style="position:absolute; left:-9999px;" alt="" /></div></noscript>

<!-- Google Analytics -->
  <script>
    window.ga = function () { ga.q.push(arguments) }; ga.q = []; ga.l = +new Date;
    ga('create', 'UA-134931760-1', 'auto'); ga('send', 'pageview')
  </script>
  <script src="https://www.google-analytics.com/analytics.js" async defer></script>

</section>

  <footer class="page-footer">
    <div class="page-footer-legal-info-container page-footer-element">
      <p>
        Weekly-Geekly | <span class="page-footer-legal-info-year js-page-footer-legal-info-year">2019</span>
      </p>
    </div>
    <div class="page-footer-counters-container page-footer-element">
      <a class="page-footer-counter-clustrmap" href='#'  title='Visit tracker'><img src='https://clustrmaps.com/map_v2.png?cl=698e5a&w=271&t=t&d=6iCFw7uJz0zcOaoxz5k5PcLCJUzv2WG8G5V8M3U6Rc4&co=3a3a3a&ct=ffffff'/></a>
    </div>
  </footer>
</body>

</html>